elytron/ELY-816-masked-password.adoc

```diff
@@ -0,0 +1,84 @@
+= ELY-816 Support for masked passwords in client XML config
+:author:            Jan Kalina
+:email:             jkalina@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+The Elytron client should support masked passwords - passwords encrypted using a predefined key, as it is supported in
+PicketBox. It should be possible to use masked passwords in place of any clear password written in the Elytron Client XML
+configuration.
+
+For example, instead of:
+[source,xml]
+----
+<credentials>
+    <clear-password password="password"/>
+</credentials>
+----
+
+The following would also be possible to use:
+
+[source,xml]
+----
+<credentials>
+    <masked-password iteration-count="12" salt="12345678" masked-password="j37uUs8kG9t2QSWjoxxtDg=="/>
+</credentials>
+----
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/ELY-816[ELY-816]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1065[EAP7-1065]
+* https://issues.jboss.org/browse/ELY-867[ELY-867]
+
+=== Dev Contacts
+
+*  mailto:aabdelsa@redhat.com[Ashley Abdel-Sayed]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* WildFly Elytron
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* Everywhere in the Elytron client configuration where a clear password can be used, it should be possible to use a masked
+password instead:
+** credential in authentication-configuration
+** key-store password
+** key-store reference (like in key-store-ssl-certificate)
+* It should be possible to migrate passwords in PicketBox-format into Elytron
+* It should be possible to represent the password in modular crypt format
+* This will require not using the initialization vectors (IV) randomly generated by the javax.crypto.Cipher as that makes
+decryption impossible
+* This will also deprecate the PKDBF2 algorithms for masked passwords as these algorithms cannot be used to get a Cipher
+instance.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+== Test Plan
+
+* Tests will be added to test masked passwords with a clear spec as well as encryptable spec (in modular crypt format).
+* Elytron XML parser tests will be added.
+* Tests will be added to test PicketBox compatability.
+* Tests using Elytron client with clear-password will be copied and implemented with masked password instead.
```

Comment:
@ashley-abdelsayed98 Please also add a "Testing By" section (we can leave this as TBD for now) and a "Community Documentation" section (since the original analysis document was written a little while ago, it's missing these sections).

Reply:
Added